File: sorting/merge.cpp
#include <iostream>
using namespace std;

void merge(int* a, int start, int mid, int end) {
  int* b = new int[end - start];

  // Indices for left and right arrays
  int l = start;
  int r = mid;

  // Index for merged array
  int i = 0;

  while(l < mid && r < end) {
    if(a[l] <= a[r]) {
      b[i] = a[l];
      l++;
    }
    else if(a[r] < a[l]) {
      b[i] = a[r];
      r++;
    }

    i++;
  }

  // Add the remaining values to the end of the merged array
  while(l < mid) {
    b[i] = a[l];
    l++;
    i++;
  }

  while(r < end) {
    b[i] = a[r];
    r++;
    i++;
  }

  // Copy the merged array into the original array
  for(i = 0; i < end - start; i++)
    a[i + start] = b[i];
}

void mergeSort(int* a, int start, int end) {
  if(end == start + 1)
    return;

  int mid = (start + end) / 2;

  mergeSort(a, start, mid);
  mergeSort(a, mid, end);
  merge(a, start, mid, end);
}

int main() {
  int length;
  cout << "Enter the number of values to be sorted:\n";
  cin >> length;

  int a[length];

  for(int i = 0; i < length; ++i) {
    cout << "Number " << (i + 1) << ":\n";
    cin >> a[i];
  }

  mergeSort(a, 0, length);

  // Display sorted array
  for(int i = 0; i < length; ++i)
    cout << a[i] << " ";
  cout << "\n";

  return 0;
}

File: sorting/merge.py
def merge(left, right):
	# Merged array
	a = []

	# Iterators for left and right arrays
	l = 0
	r = 0

	while l < len(left) and r < len(right):
		if left[l] <= right[r]:
			a.append(left[l])
			l += 1
		elif right[r] < left[l]:
			a.append(right[r])
			r += 1

	# Add remaining values to the merged array
	a.extend(left[l:])
	a.extend(right[r:])

	return a

def merge_sort(a):
	if len(a) == 1:
		return a

	mid = len(a) / 2

	left = merge_sort(a[:mid])
	right = merge_sort(a[mid:])
	return merge(left, right)

a = []
length = input('Please enter the number of values you wish to sort: ')

for i in range(0, length):
	a.append(input('Number %s: ' % (i + 1)))

a = merge_sort(a)
print a